statsd/dispatcher.go

```diff
@@ -0,0 +1,185 @@
+package statsd
+
+import (
+	"hash/adler32"
+	"sync"
+	"time"
+
+	"github.com/atlassian/gostatsd/types"
+
+	log "github.com/Sirupsen/logrus"
+	"golang.org/x/net/context"
+)
+
+// Dispatcher is responsible for managing Aggregators' lifecycle and dispatching metrics among them.
+type Dispatcher interface {
+	Run(context.Context) error
+	DispatchMetric(context.Context, *types.Metric) error
+	Flush(context.Context) <-chan *types.MetricMap
+	Process(context.Context, ProcessFunc) *sync.WaitGroup
+}
+
+// AggregatorFactory creates Aggregator objects.
+type AggregatorFactory interface {
+	// Create creates Aggregator objects.
+	Create() Aggregator
+}
+
+// AggregatorFactoryFunc type is an adapter to allow the use of ordinary functions as AggregatorFactory.
+type AggregatorFactoryFunc func() Aggregator
+
+// Create calls f().
+func (f AggregatorFactoryFunc) Create() Aggregator {
+	return f()
+}
+
+type flushCommand struct {
+	ctx    context.Context
+	wg     sync.WaitGroup
+	result chan<- *types.MetricMap
+}
+
+type processCommand struct {
+	wg sync.WaitGroup
+	f  ProcessFunc
+}
+
+type worker struct {
+	metrics     chan *types.Metric
+	flushChan   chan *flushCommand
+	processChan chan *processCommand
+	aggr        Aggregator
+}
+
```

Comment:
We could. I'd prefer to keep it as a struct to be able to easily add fields later if needed - making it a type alias does not buy us much in this case.